### DISCUSSION TITLE
Add simple mock code to dev mode

### DISCUSSION THREAD

**Discussion by t1** [quarkusio/quarkus#43745] [Q&A] [answered]
I'd like to add a few classes while running in dev-mode (actually also for tests, but not in my prod code) that mock the API of a third party service. They could listen on a separate port, but it would also be okay to run them on the same port as my application.

I understand that I can run the WireMock dev-service, but that's not stateful. Even when I try to add a WireMock extension to supply my own stub mappings from code, the dev-service doesn't load the classes from `src/test/java`, only the `src/test/resources/mappings`.

Starting a separate Quarkus maven module might be a solution, but I would have to work with fixed ports; and running it in my CD pipeline is not really trivial.

Are there better options?

**Comment by donkon** [comment, marked as answer]
I faced the same problem.

I solved it by compiling / adding the mocks (located in src/test/java) to the classpath by using the [build-helper:add-source](https://www.mojohaus.org/build-helper-maven-plugin/add-source-mojo.html) plugin.

```
@Alternative
@Priority(1)
@ApplicationScoped
public class MyMock extends MyService {

}
```
I added following profile to my pom.xml:
```
<profile>
	<id>dev</id>
	<build>
		<plugins>
			<plugin>
				<groupId>org.codehaus.mojo</groupId>
				<artifactId>build-helper-maven-plugin</artifactId>
				<version>3.0.0</version>
				<executions>
					<execution>
						<phase>generate-sources</phase>
						<goals>
							<goal>add-source</goal>
						</goals>
						<configuration>
							<sources>
								<source>src/test/java</source>
							</sources>
						</configuration>
					</execution>
				</executions>
			</plugin>
                        <plugin>
				<artifactId>maven-compiler-plugin</artifactId>
				<version>${compiler-plugin.version}</version>
				<configuration>
					<parameters>true</parameters>
					<excludes>
						<exclude>**/*Test.java</exclude>
					</excludes>
				</configuration>
			</plugin>
		</plugins>
	</build>
</profile>
```
By running `./mvnw quarkus:dev -P dev` the mocks are now on the classpath and get injected instead of the real service (note that `@Mock` does not work in this case).

**Reply by t1** [reply]
@donkon: That's actually a nice idea. Thanks for sharing! I was hoping for something more... idiomatic. But I'll give your approach a try.

**Reply by aloubyansky** [reply]
You could also add  `<defaultGoal>quarkus:dev</defaultGoal>` under `<build>`  and then launch with just `mvn -Pdev`

**Reply by t1** [reply]
I tried it and it works really well. I came up with a variation that I personally like better, but that's probably just me: I put my mocks into a new directory `src/dev/java` for the tests:

```xml
            <plugin>
                <groupId>org.codehaus.mojo</groupId>
                <artifactId>build-helper-maven-plugin</artifactId>
                <executions>
                    <execution>
                        <phase>generate-sources</phase>
                        <goals>
                            <goal>add-test-source</goal>
                        </goals>
                        <configuration>
                            <sources>
                                <source>src/dev/java</source>
                            </sources>
                        </configuration>
                    </execution>
                </executions>
            </plugin>
```

And add it to the normal resources in the `dev` profile:

```xml
        <profile>
            <id>dev</id>
            <properties>
                <!-- enable debug port for remote debugging -->
                <debug>5005</debug>
            </properties>
            <build>
                <plugins>
                    <plugin>
                        <groupId>org.codehaus.mojo</groupId>
                        <artifactId>build-helper-maven-plugin</artifactId>
                        <executions>
                            <execution>
                                <phase>generate-sources</phase>
                                <goals>
                                    <goal>add-source</goal>
                                </goals>
                                <configuration>
                                    <sources>
                                        <source>src/dev/java</source>
                                    </sources>
                                </configuration>
                            </execution>
                        </executions>
                    </plugin>
                </plugins>
            </build>
        </profile>
```

**Comment by phillip-kruger** [comment]
@aloubyansky can your proposed dev only dependency help here ?

**Reply by aloubyansky** [reply]
This is something else. Then approach I am looking into is enabling an extension to declare a dev mode dependency on another extension. We could also support an extension depending in dev mode on any artifact, not just an extension. If that's useful.

**Reply by t1** [reply]
That sounds interesting, too!
